halo2_middleware/src/zal.rs

```diff
@@ -0,0 +1,233 @@
+//! - Pointer(s) caching to limit data movement between CPU and GPU, FPGAs
+//! - Length of data
+//! - data in layout:
+//!    - canonical or Montgomery fields, unsaturated representation, endianness
+//!    - jacobian or projective coordinates or maybe even Twisted Edwards for faster elliptic curve additions,
+//!    - FFT: canonical or bit-reversed permuted
+//! - data out layout
+//! - Device(s) ID
+//!
+//! They are required to be Plain Old Data (Copy trait), so no custom `Drop` is required.
+//! If a specific resource is needed, it can be stored in the engine in a hashmap for example
+//! and an integer ID or a pointer can be opaquely given as a descriptor.
+
+// The ZK Accel Layer API
+// ---------------------------------------------------
+pub mod traits {
+    use halo2curves::CurveAffine;
+
+    pub trait MsmAccel<C: CurveAffine> {
+        fn msm(&self, coeffs: &[C::Scalar], base: &[C]) -> C::Curve;
+
+        // Caching API
+        // -------------------------------------------------
+        // From here we propose an extended API
+        // that allows reusing coeffs and/or the base points
+        //
+        // This is inspired by CuDNN API (Nvidia GPU)
+        // and oneDNN API (CPU, OpenCL) https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnn-ops-infer-so-opaque
+        // usage of descriptors
+        //
+        // https://github.com/oneapi-src/oneDNN/blob/master/doc/programming_model/basic_concepts.md
+        //
+        // Descriptors are opaque pointers that hold the input in a format suitable for the accelerator engine.
+        // They may be:
+        // - Input moved on accelerator device (only once for repeated calls)
+        // - Endianess conversion
+        // - Converting from Montgomery to Canonical form
+        // - Input changed from Projective to Jacobian coordinates or even to a Twisted Edwards curve.
+        // - other form of expensive preprocessing
+        type CoeffsDescriptor<'c>: Copy;
+        type BaseDescriptor<'b>: Copy;
+
+        fn get_coeffs_descriptor<'c>(&self, coeffs: &'c [C::Scalar]) -> Self::CoeffsDescriptor<'c>;
+        fn get_base_descriptor<'b>(&self, base: &'b [C]) -> Self::BaseDescriptor<'b>;
+
+        fn msm_with_cached_scalars(
+            &self,
+            coeffs: &Self::CoeffsDescriptor<'_>,
+            base: &[C],
+        ) -> C::Curve;
+
+        fn msm_with_cached_base(
+            &self,
+            coeffs: &[C::Scalar],
+            base: &Self::BaseDescriptor<'_>,
+        ) -> C::Curve;
+
+        fn msm_with_cached_inputs(
+            &self,
+            coeffs: &Self::CoeffsDescriptor<'_>,
+            base: &Self::BaseDescriptor<'_>,
+        ) -> C::Curve;
+        // Execute MSM according to descriptors
+        // Unsure of naming, msm_with_cached_inputs, msm_apply, msm_cached, msm_with_descriptors, ...
+    }
+}
+
+// ZAL using Halo2curves as a backend
+// ---------------------------------------------------
+
+pub mod impls {
```

Comment:
Ideally this implementation should live in halo2curves.  Nevertheless doing this leads to a cyclic dependency because:
- `halo2_middleware` imports `halo2curves` for the `CurveAffine` trait
- `halo2curves` imports `halo2_middleware` for the `MsmAccel` trait

Until `CurveAffine` is moved to the `group` crate we can't resolve this cleanly, but we can do the following workaround:
- `CurveAffine` is currently re-exported by `halo2curves` from `pasta`
- We can always import `CurveAffine` from `pasta` everywhere to break the cycle
    - In particular, in `halo2_middleware` we can import `CurveAffine` from `pasta` so that `halo2curves` can import `halo2_middleware` without introducing a cycle.

Reply:
> In particular, in halo2_middleware we can import CurveAffine from pasta so that halo2curves can import halo2_middleware without introducing a cycle.

I think that's OK if they are in the same workspace / monorepo, but there is a hidden/social assumption that `halo2` depends on `halo2curves` and NOT vice-versa.

That said, should I do anything here or do we leave that to a future RFC or PR?

Reply:
> That said, should I do anything here or do we leave that to a future RFC or PR?

I'm inclined to leave this for a future PR so that we can make progress with testing ZAL.

In particular, we need to figure out what would be the ideal hierarchy of crates; and then based the current constraints (the fact that we don't control `ff`, `group`, `pasta`), figure out what's the best we can do; but I think that may take some time and discussion and it shouldn't block the first iteration of ZAL.